pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,288 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/pkg/antctl/raw/multicluster/common"
+)
+
+type JoinOptions struct {
+	Kind            string     `yaml:"kind"`
+	APIVersion      string     `yaml:"apiVersion"`
+	ClusterSetID    string     `yaml:"clusterSetID"`
+	ClusterID       string     `yaml:"clusterID"`
+	Namespace       string     `yaml:"namespace,omitempty"`
+	LeaderClusterID string     `yaml:"leaderClusterID"`
+	LeaderNamespace string     `yaml:"leaderNamespace"`
+	LeaderAPIServer string     `yaml:"leaderAPIServer"`
+	TokenSecretName string     `yaml:"tokenSecretName,omitempty"`
+	TokenSecretFile string     `yaml:"tokenSecretFile,omitempty"`
+	ConfigFile      string     `yaml:"configFile,omitempty"`
+	Secret          *v1.Secret `yaml:"-"`
+}
+
+var joinOpts *JoinOptions
+
+func (o *JoinOptions) validateAndComplete() error {
+	if o.ConfigFile != "" {
+		raw, err := ioutil.ReadFile(o.ConfigFile)
+		if err != nil {
+			return err
+		}
+		if err := yamlUnmarshall(raw, o); err != nil {
+			return err
+		}
+		if o.Kind != common.ClusterSetJoinConfigKind || o.APIVersion != common.ClusterSetJoinConfigAPIVersion {
+			return fmt.Errorf("unknown apiVersion or kind in config: %s", o.ConfigFile)
+		}
+
+		o.Secret, err = unmarshallSecret(raw, true)
```

Comment:
If we believe TokenSecretFile and TokenSecretName in the config file should take precedence than the default token manifest, we should check if they are set or not first. If we keep your current strategy, we should remove TokenSecretFile from ClusterSetJoinConfig or at least remove it from YAML fields.

Reply:
Question - what if the config file does not include a Secret manifest? We should not fail here, as long as JoinOptions includes one of TokenSecretName or TokenSecretFile right?

Reply:
Fixed. Changed the Secret order now. 
The order of the Secret is tokenSecretName > tokenSecretFile > joinConfigFile. So if we need to unmarshall the Secret from the Config file. It means that we do not set the `tokenSecretName` and `tokenSecretFile`.  So we need to throw an error here.